### ISSUE TITLE
Document how word2vec is training

### ISSUE THREAD

**Issue by Make42** [RaRe-Technologies/gensim#1920]
I would like to see a documentation how word2vec is training skip-gram and CBOW.

### First, the Skip-Gram model:

From the Internet I found two alternative ways. For that let us establish that each word is encoded as a one-hot-vector, which is put into the NN returning a probability vector for each word of finding this word near (in the neighborhood, being a context word) the input word. Let us assume there are `C` context words for each word, which we can also one-hot-encode. For training we have a word and it's `C` context words as targets (all vectors one-hot-encoded).

1. https://www.analyticsvidhya.com/blog/2017/06/word-embeddings-count-word2veec/ describes that the NN output of a word is compared with each of the `C` target context words one-hot-vectors. For each of the comparisons an error vector is calculated. Those error vectors are added up to the total error vector and this is used for the training / updating of the weights via backpropagation. So the a training set is `Word -> {C1, C2, C3, C4}`. This is also the description of https://iksinc.online/tag/continuous-bag-of-words-cbow/ and I get the impression it is also what the original publication is suggesting.
2. http://mccormickml.com/2016/04/19/word2vec-tutorial-the-skip-gram-model/ suggests each word produces up to `C` training sets (less than `C` if the word is at the beginning or end of the sentence). So one word produces multiple training sets for backpropagation, so `Word -> C1, Word -> C2, Word -> C3, Word -> C4` and for each backpropagation there is only one error vector calculated.

To find out how gensim is doing it checked `gensim.models.word2vec` and there especially the method

```
    def train_batch_sg(model, sentences, alpha, work=None, compute_loss=False):
        result = 0
        for sentence in sentences:
            word_vocabs = [model.wv.vocab[w] for w in sentence if w in model.wv.vocab and
                           model.wv.vocab[w].sample_int > model.random.rand() * 2**32]
            for pos, word in enumerate(word_vocabs):
                reduced_window = model.random.randint(model.window)  # `b` in the original word2vec code

                # now go over all words from the (reduced) window, predicting each one in turn
                start = max(0, pos - model.window + reduced_window)
                for pos2, word2 in enumerate(word_vocabs[start:(pos + model.window + 1 - reduced_window)], start):
                    # don't train on the `word` itself
                    if pos2 != pos:
                        train_sg_pair(
                            model, model.wv.index2word[word.index], word2.index, alpha, compute_loss=compute_loss
                        )

            result += len(word_vocabs)
        return result
```

from which I get the strong impression, that 2. way is used.

### Next, the CBOW model:

1. https://www.analyticsvidhya.com/blog/2017/06/word-embeddings-count-word2veec/ describes that each context word is put through the NN up to the hidden layer. There the results (after the hidden layer) of the different `C` context words are aggregated (using the mean). The aggregate is put through the output layer and the output is calculated. This is compared with the target word vector (one-hot-encoded), which is the "central" word of the sliding window that goes over the sentences. The resulting error vector is used for updating the weights via backpropgation. This is also the description of https://iksinc.online/tag/continuous-bag-of-words-cbow/ and I get the impression it is also what the original publication is suggesting.
2. An alternative (no reference) would be to also build CBOW analogous to 2. for the Skip-Gram training.

To find out how gensim is doing it checked `gensim.models.word2vec` and there especially the method
`train_cbow_pair` which is, with

```
    [...]
    if learn_vectors:
        [...]
            if not model.cbow_mean and input_word_indices:
                neu1e /= len(input_word_indices)
            for i in input_word_indices:
                context_vectors[i] += neu1e * context_locks[i]
```

suggesting that 1. is used.

### How is gensim doing it?

Please document how gensim is doing it. If my assessment is right and you want to, please feel free to build on my descriptions.

If I am right though, I found find it strange, that the skip-gram is not following the publication and I would appreciate an explanation.

**Comment by menshikh-iv** [comment]
CC @gojomo

**Comment by menshikh-iv** [comment]
@Make42 hello, I agree, good point.
We already have it in plans (some additional pages with the description "How model X works")!

This will be your main task soon @steremma :)

**Comment by steremma** [comment]
Sounds interesting! I have already started looking at the papers introducing both methods from Tomas Mikolov. I am wrapping up with some tasks and will hopefully dive into this next week.

**Comment by gojomo** [comment]
The source is the ultimate documentation of how gensim is doing its calculations – and the gensim source was initially closely modeled after Google's `word2vec.c` implementation. So it terms of the micro-decisions about how loops are set up, and skipgram micro-training-examples (input_word -> predicted_word) are chosen, it closely matches that original reference implementation. 

In particular, for skipgram. you can either think of the neural network as learning how to predict a single target central word, from each of its surrounding context-words, OR as learning how to predict each target surrounding context word, from a single central word. Both ways result in the exact same set of (input_word -> predicted_word) skipgram pairs, just in a slightly different order. IIUC, the original paper described the loop one way, but then Google's implementors found the other way ran slightly faster (likely due to CPU cache issues), so wrote `word2vec.c` the alternate way, which gensim then copied. 

@Make42 - your description of CBOW leaves out any mention of the essential *averaging* of vectors for all words in the context window. That single averaged vector is what is then used as the input for the target-word prediction NN. Then, the error the NN discovers for that single averaged vector is also applied to each word-vector that contributed to the average. (It is the same in the paper, `word2vec.c`, and gensim.)

**Comment by Make42** [comment]
@gojomo - since by Python is better than my C, I read the gensim source code instead of the word2vec.c implementation ;-).

Regarding CBOW: https://www.analyticsvidhya.com/blog/2017/06/word-embeddings-count-word2veec/  is doing the averaging via mean after the hidden layer. My description was wrong or at least misleading - I am going to correct it.

Regarding Skip-Gram: As far as I understand I have to disagree: Skip-Gram is taking in one one-hot-encoded word and outputs the probability of each dictionary word that it is one of the surrounding words. CBOW takes in multiple words an predicts the central word. Skip-Gram and CBOW are also different in their results regarding the word embedding. Whether the implementation in word2vec.c is like 1. or 2. in my description at the top post, I do not know. This is my question.

**Comment by gojomo** [comment]
From a quick glance, neither (1) nor (2) seems to me an accurate summary of what the code is doing. Perhaps (2) is closer, because each skipgram pair (input_word -> predicted_word) is treated as an individual training example for the NN, and its backprop is finished before the next (input_word -> predicted_word) pair is considered. 

But still, to understand the gensim code, I'd recommend reviewing it directly - and not trying to map it into other summaries/explanations which may themselves be rough/incomplete. (And as previously mentioned, the gensim code very closely follows the `word2vec.c` code that came from the original Word2Vec paper authors – so if it's deviating from implications of the original paper any, it's doing so in the same way as the paper's authors themselves chose when implementing.)

**Comment by Make42** [comment]
@gojomo - In general I more interested in the mathematical setup than which data collection is used. If the input is a one-hot-encoded vector, the dot product with a matrix, the parsing through a NN with linear neurons or the lookup in a dictionary containing the dense vectors as values are effectively all the same operations. The updating of the matrix / NN weights / dense vectors still needs to happen.

I am aware of the negative sampling :-), but still thanks for pointing it out!

The bottom line is "review the code" - sure, but  the code could be more clear, even if one has a mathematical background - no offense :-). E.g. the names of the variables are not very clear, just saying `word`, `word2` or `neu1e`.

**Comment by piskvorky** [comment]
@Make42 Some toy implementation of word2vec may be a better place for your studies. The Gensim implementation is highly optimized, plus it has been battle-hardened and extended by several additional functionalities (like scoring, newer models, parallelization) over the years.

**Comment by Huixxi** [comment]
I have wirte a skip-gram model with negative sampling method as the optimize method.Here is my [code](https://github.com/Huixxi/NLP_Word2Vec/blob/master/bin/word2vec_test.py).I have read a lot tuturals like blogs or papers and also checked several times but couldn't find any wrong in the structure or logic of my model, the performence of my model is always far worse than the gensim skip-gram version, only get about 3% accuracy on that test dataset. So I think maybe I can get some help here. Thanks.

**Comment by AMR-KELEG** [comment]
I agree with @Make42 that this part causes high confusion.
I generally try to read some introductory blogs and compare them to the implementation.
I didn't really find a blog that mentions the current implementation of gensim for the skipgram model which left me with big confusions especially on checking the shape of the weight matrices.
